File: zroky-sdk/zroky/_verified_action.py
# SPDX-License-Identifier: FSL-1.1-MIT
# Copyright 2026 Zroky AI

"""Verified action client for backend-owned execution, verification, and receipt."""
from __future__ import annotations

import time
from datetime import datetime
from typing import Any, Mapping
from urllib.parse import urlsplit, urlunsplit
from uuid import uuid4

import httpx

from zroky._errors import (
    ZrokyVerifiedActionApprovalRequired,
    ZrokyVerifiedActionBlocked,
    ZrokyVerifiedActionError,
)

_REQUEST_TIMEOUT_S = 8.0
_TERMINAL_PROOF_STATUSES = {"matched", "mismatched", "not_verified"}
_TERMINAL_RECEIPT_STATUSES = {"generated", "failed"}

_EXECUTION_REQUEST_FORBIDDEN_KEYS = {
    "runner_id",
    "runner",
    "credential_ref",
    "credential_reference",
    "protected_credential_ref",
}
_EXECUTION_REQUEST_RAW_SECRET_KEYS = {
    "authorization",
    "bearer_token",
    "api_key",
    "apikey",
    "password",
    "secret",
    "token",
}
_EXECUTION_REQUEST_RAW_SECRET_VALUES = (
    "bearer ",
    "sk_live_",
    "sk_test_",
    "xoxb-",
    "xoxp-",
    "ghp_",
    "gho_",
    "github_pat_",
    "-----begin private key-----",
)


def _api_base(ingest_url: str) -> str:
    parsed = urlsplit(ingest_url)
    if not parsed.scheme or not parsed.netloc:
        return ingest_url.rstrip("/")

    path = parsed.path.rstrip("/")
    for suffix in ("/api/v1/ingest", "/v1/ingest", "/ingest"):
        if path.endswith(suffix):
            path = path[: -len(suffix)].rstrip("/")
            break
    return urlunsplit((parsed.scheme, parsed.netloc, path, "", "")).rstrip("/")


def _api_url(ingest_url: str, path: str) -> str:
    return f"{_api_base(ingest_url)}{path}"


def _execution_request_violation(value: Any, *, key_path: tuple[str, ...] = ()) -> str | None:
    if isinstance(value, Mapping):
        for key, nested in value.items():
            key_text = str(key).strip().lower()
            if key_text in _EXECUTION_REQUEST_FORBIDDEN_KEYS:
                return ".".join((*key_path, str(key)))
            if any(marker in key_text for marker in _EXECUTION_REQUEST_RAW_SECRET_KEYS):
                return ".".join((*key_path, str(key)))
            found = _execution_request_violation(nested, key_path=(*key_path, str(key)))
            if found is not None:
                return found
        return None
    if isinstance(value, (list, tuple)):
        for index, nested in enumerate(value):
            found = _execution_request_violation(nested, key_path=(*key_path, str(index)))
            if found is not None:
                return found
        return None
    if isinstance(value, str):
        lowered = value.strip().lower()
        if any(marker in lowered for marker in _EXECUTION_REQUEST_RAW_SECRET_VALUES):
            return ".".join(key_path) or "execution_request"
    return None


def _validate_execution_request(execution_request: Mapping[str, Any] | None) -> dict[str, Any] | None:
    if execution_request is None:
        return None
    if not isinstance(execution_request, Mapping):
        raise ZrokyVerifiedActionError("[ZROKY] execution_request must be an object.")
    execution_plan = execution_request.get("execution_plan")
    if not isinstance(execution_plan, Mapping) or not execution_plan:
        raise ZrokyVerifiedActionError("[ZROKY] execution_request.execution_plan must be a non-empty object.")
    credential_pointer = execution_request.get("credential_pointer")
    if isinstance(credential_pointer, str) and "://" in credential_pointer:
        raise ZrokyVerifiedActionError("[ZROKY] execution_request.credential_pointer must be a non-secret alias.")
    credential = execution_request.get("credential")
    if isinstance(credential, Mapping):
        nested_pointer = credential.get("pointer")
        if isinstance(nested_pointer, str) and "://" in nested_pointer:
            raise ZrokyVerifiedActionError("[ZROKY] execution_request.credential.pointer must be a non-secret alias.")
    violation = _execution_request_violation(execution_request)
    if violation is not None:
        raise ZrokyVerifiedActionError(
            "[ZROKY] execution_request must not include runner pins, protected "
            f"credential refs, or raw secret material at {violation}."
        )
    return dict(execution_request)


def _ensure_config():
    import zroky as _z  # lazy import keeps global config in sync

    if _z._config is None:
        _z.init()
    cfg = _z._config
    if cfg is None or not cfg.api_key or not cfg.project:
        raise ZrokyVerifiedActionError("[ZROKY] verified_action requires api_key and project.")
    return cfg


def _headers(cfg: Any, *, idempotency_key: str | None = None) -> dict[str, str]:
    headers = {
        "Content-Type": "application/json",
        "x-api-key": cfg.api_key,
        "x-project-id": cfg.project,
        "Authorization": f"Bearer {cfg.api_key}",
    }
    if idempotency_key:
        headers["Idempotency-Key"] = idempotency_key
    return headers


def _request_json(
    method: str,
    path: str,
    *,
    json_payload: dict[str, Any] | None = None,
    idempotency_key: str | None = None,
    timeout: float = _REQUEST_TIMEOUT_S,
) -> dict[str, Any]:
    cfg = _ensure_config()
    try:
        response = httpx.request(
            method,
            _api_url(cfg.ingest_url, path),
            headers=_headers(cfg, idempotency_key=idempotency_key),
            json=json_payload,
            timeout=timeout,
        )
    except httpx.HTTPError as exc:
        raise ZrokyVerifiedActionError(f"[ZROKY] verified action API unavailable: {exc}") from exc

    if response.status_code >= 300:
        raise ZrokyVerifiedActionError(
            f"[ZROKY] verified action API failed with HTTP {response.status_code}: {response.text[:240]}"
        )
    try:
        data = response.json()
    except ValueError as exc:
        raise ZrokyVerifiedActionError("[ZROKY] verified action API returned invalid JSON.") from exc
    if not isinstance(data, dict):
        raise ZrokyVerifiedActionError("[ZROKY] verified action API returned a non-object response.")
    return data


def _without_none(value: dict[str, Any]) -> dict[str, Any]:
    return {key: item for key, item in value.items() if item is not None}


def _iso_or_value(value: datetime | str | None) -> str | None:
    if isinstance(value, datetime):
        return value.isoformat()
    return value


def _requires_approval(decision: dict[str, Any]) -> bool:
    return bool(decision.get("requires_approval")) or decision.get("status") == "approval_pending"


def _blocked(decision: dict[str, Any]) -> bool:
    return bool(decision.get("allowed")) is False and not _requires_approval(decision)


def verified_action(
    *,
    agent_id: str | None = None,
    contract_version: str,
    action_type: str,
    operation_kind: str,
    environment: str = "production",
    principal: dict[str, Any] | None = None,
    actor_chain: list[dict[str, Any]] | None = None,
    purpose: dict[str, Any] | None = None,
    resource: dict[str, Any] | None = None,
    parameters: dict[str, Any] | None = None,
    execution_request: Mapping[str, Any] | None = None,
    verification_profile: str | None = None,
    deadline: datetime | str | None = None,
    trace_context: dict[str, Any] | None = None,
    idempotency_key: str | None = None,
    raise_on_approval: bool = True,
) -> dict[str, Any]:
    """Create and authorize a backend-owned verified action.

    The agent supplies only intent plus an execution_request. It must not name a
    runner or protected credential ref; the backend resolves those from runner
    configuration, then workers execute and verify independently.
    """

    checked_execution_request = _validate_execution_request(execution_request)
    effective_idempotency_key = idempotency_key or f"zroky-sdk:{uuid4()}"
    cfg = _ensure_config()
    payload = _without_none(
        {
            "agent_id": agent_id or cfg.default_agent_id,
            "contract_version": contract_version,
            "action_type": action_type,
            "operation_kind": operation_kind,
            "environment": environment,
            "principal": principal or {},
            "actor_chain": actor_chain or [],
            "purpose": purpose or {},
            "resource": resource or {},
            "parameters": parameters or {},
            "execution_request": checked_execution_request,
            "verification_profile": verification_profile,
            "deadline": _iso_or_value(deadline),
            "trace_context": trace_context,
        }
    )

    action = _request_json(
        "POST",
        "/v1/action-intents",
        json_payload=payload,
        idempotency_key=effective_idempotency_key,
    )
    action_id = str(action.get("action_id") or "")
    if not action_id:
        raise ZrokyVerifiedActionError("[ZROKY] verified action create response did not include action_id.")

    decision = _request_json("POST", f"/v1/action-intents/{action_id}/decide", json_payload={})
    if _requires_approval(decision):
        if raise_on_approval:
            raise ZrokyVerifiedActionApprovalRequired(
                "[ZROKY] verified action requires approval before execution.",
                action=action,
                decision=decision,
            )
        return decision
    if _blocked(decision):
        raise ZrokyVerifiedActionBlocked(
            "[ZROKY] verified action was blocked by policy.",
            action=action,
            decision=decision,
        )
    return decision


def await_action_proof(
    action_id: str,
    *,
    timeout_seconds: float = 120.0,
    poll_interval_seconds: float = 2.0,
) -> dict[str, Any]:
    """Poll until backend verification and receipt generation reach a terminal state."""

    deadline = time.monotonic() + max(0.1, timeout_seconds)
    last_action: dict[str, Any] | None = None
    while time.monotonic() < deadline:
        action = _request_json("GET", f"/v1/action-intents/{action_id}")
        last_action = action
        proof_status = str(action.get("proof_status") or "")
        receipt_status = str(action.get("receipt_status") or "")
        if proof_status in _TERMINAL_PROOF_STATUSES and receipt_status in _TERMINAL_RECEIPT_STATUSES:
            receipt: dict[str, Any] | None = None
            if receipt_status == "generated":
                receipt = _request_json("GET", f"/v1/action-intents/{action_id}/receipt")
            return {
                "action_id": action_id,
                "action": action,
                "receipt": receipt,
                "proof_status": proof_status,
                "receipt_status": receipt_status,
                "signature_valid": receipt.get("signature_valid") if receipt else None,
                "evidence_id": receipt.get("receipt_id") if receipt else None,
            }
        time.sleep(max(0.05, poll_interval_seconds))

    raise ZrokyVerifiedActionError(
        "[ZROKY] timed out waiting for verified action proof.",
        action=last_action,
    )

File: zroky-sdk-js/src/verified-action.ts
// SPDX-License-Identifier: FSL-1.1-MIT
// Copyright 2026 Zroky AI

import { apiUrl, nodeEnv } from "./api";
import { resolveConfig } from "./config";
import type { ZrokyConfig } from "./types";

type JsonObject = Record<string, unknown>;

const TERMINAL_PROOF_STATUSES = new Set(["matched", "mismatched", "not_verified"]);
const TERMINAL_RECEIPT_STATUSES = new Set(["generated", "failed"]);

const FORBIDDEN_EXECUTION_REQUEST_KEYS = new Set([
  "runner_id",
  "runner",
  "credential_ref",
  "credential_reference",
  "protected_credential_ref",
]);
const RAW_SECRET_KEY_MARKERS = [
  "authorization",
  "bearer_token",
  "api_key",
  "apikey",
  "password",
  "secret",
  "token",
];
const RAW_SECRET_VALUE_MARKERS = [
  "bearer ",
  "sk_live_",
  "sk_test_",
  "xoxb-",
  "xoxp-",
  "ghp_",
  "gho_",
  "github_pat_",
  "-----begin private key-----",
];

export interface VerifiedActionOptions {
  agentId?: string;
  contractVersion: string;
  actionType: string;
  operationKind: string;
  environment?: string;
  principal?: JsonObject;
  actorChain?: JsonObject[];
  purpose?: JsonObject;
  resource?: JsonObject;
  parameters?: JsonObject;
  executionRequest?: JsonObject;
  verificationProfile?: string;
  deadline?: string | Date;
  traceContext?: JsonObject;
  idempotencyKey?: string;
  raiseOnApproval?: boolean;
}

export interface VerifiedActionDecision extends JsonObject {
  action_id?: string;
  status?: string;
  allowed?: boolean;
  requires_approval?: boolean;
  runtime_policy_decision_id?: string | null;
}

export interface AwaitActionProofOptions {
  timeoutMs?: number;
  pollIntervalMs?: number;
}

export interface ActionProofResult {
  actionId: string;
  action: JsonObject;
  receipt: JsonObject | null;
  proofStatus: string;
  receiptStatus: string;
  signatureValid: boolean | null;
  evidenceId: string | null;
}

export class ZrokyVerifiedActionError extends Error {
  action: JsonObject;
  decision: JsonObject;

  constructor(message: string, options: { action?: JsonObject; decision?: JsonObject } = {}) {
    super(message);
    this.name = "ZrokyVerifiedActionError";
    this.action = options.action ?? {};
    this.decision = options.decision ?? {};
  }
}

export class ZrokyVerifiedActionBlocked extends ZrokyVerifiedActionError {
  constructor(message: string, options: { action?: JsonObject; decision?: JsonObject } = {}) {
    super(message, options);
    this.name = "ZrokyVerifiedActionBlocked";
  }
}

export class ZrokyVerifiedActionApprovalRequired extends ZrokyVerifiedActionBlocked {
  actionId?: string;
  approvalId?: string;

  constructor(action: JsonObject, decision: JsonObject) {
    super("[ZROKY] verified action requires approval before execution.", { action, decision });
    this.name = "ZrokyVerifiedActionApprovalRequired";
    this.actionId = text(action.action_id ?? decision.action_id);
    this.approvalId = text(
      decision.runtime_policy_decision_id ??
        decision.id ??
        (isRecord(decision.approval_queue_item)
          ? decision.approval_queue_item.id
          : undefined),
    );
  }
}

function text(value: unknown): string | undefined {
  return typeof value === "string" && value.length > 0 ? value : undefined;
}

function isRecord(value: unknown): value is JsonObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function withoutUndefined(value: JsonObject): JsonObject {
  return Object.fromEntries(Object.entries(value).filter(([, item]) => item !== undefined));
}

function deadlineValue(value: string | Date | undefined): string | undefined {
  if (value instanceof Date) {
    return value.toISOString();
  }
  return value;
}

function defaultIdempotencyKey(): string {
  const maybeCrypto = globalThis.crypto as { randomUUID?: () => string } | undefined;
  const random = maybeCrypto?.randomUUID?.() ?? `${Date.now()}-${Math.random().toString(16).slice(2)}`;
  return `zroky-sdk:${random}`;
}

function findExecutionRequestViolation(value: unknown, keyPath: string[] = []): string | undefined {
  if (Array.isArray(value)) {
    for (let index = 0; index < value.length; index += 1) {
      const found = findExecutionRequestViolation(value[index], [...keyPath, String(index)]);
      if (found) return found;
    }
    return undefined;
  }
  if (isRecord(value)) {
    for (const [key, nested] of Object.entries(value)) {
      const keyText = key.trim().toLowerCase();
      if (FORBIDDEN_EXECUTION_REQUEST_KEYS.has(keyText)) {
        return [...keyPath, key].join(".");
      }
      if (RAW_SECRET_KEY_MARKERS.some((marker) => keyText.includes(marker))) {
        return [...keyPath, key].join(".");
      }
      const found = findExecutionRequestViolation(nested, [...keyPath, key]);
      if (found) return found;
    }
    return undefined;
  }
  if (typeof value === "string") {
    const lowered = value.trim().toLowerCase();
    if (RAW_SECRET_VALUE_MARKERS.some((marker) => lowered.includes(marker))) {
      return keyPath.join(".") || "execution_request";
    }
  }
  return undefined;
}

function validateExecutionRequest(value: JsonObject | undefined): JsonObject | undefined {
  if (value === undefined) {
    return undefined;
  }
  const { executionPlan, credentialPointer, ...rest } = value;
  const normalized: JsonObject = {
    ...rest,
    execution_plan: value.execution_plan ?? executionPlan,
    credential_pointer: value.credential_pointer ?? credentialPointer,
  };
  if (!isRecord(normalized.execution_plan) || Object.keys(normalized.execution_plan).length === 0) {
    throw new ZrokyVerifiedActionError("[ZROKY] executionRequest.executionPlan must be a non-empty object.");
  }
  if (typeof normalized.credential_pointer === "string" && normalized.credential_pointer.includes("://")) {
    throw new ZrokyVerifiedActionError("[ZROKY] executionRequest.credentialPointer must be a non-secret alias.");
  }
  if (isRecord(normalized.credential) && typeof normalized.credential.pointer === "string" && normalized.credential.pointer.includes("://")) {
    throw new ZrokyVerifiedActionError("[ZROKY] executionRequest.credential.pointer must be a non-secret alias.");
  }
  const violation = findExecutionRequestViolation(normalized);
  if (violation) {
    throw new ZrokyVerifiedActionError(
      `[ZROKY] executionRequest must not include runner pins, protected credential refs, or raw secret material at ${violation}.`,
    );
  }
  return withoutUndefined(normalized);
}

function credentials(config: ZrokyConfig): { apiKey: string; projectId: string; endpoint?: string } {
  const env = nodeEnv();
  const apiKey = config.apiKey ?? env?.["ZROKY_API_KEY"];
  const projectId = config.projectId ?? env?.["ZROKY_PROJECT_ID"];
  if (!apiKey || !projectId) {
    throw new ZrokyVerifiedActionError("[ZROKY] verifiedAction requires apiKey and projectId.");
  }
  return { apiKey, projectId, endpoint: config.endpoint ?? env?.["ZROKY_ENDPOINT"] };
}

async function requestJson(
  method: string,
  path: string,
  config: ZrokyConfig,
  options: { body?: JsonObject; idempotencyKey?: string } = {},
): Promise<JsonObject> {
  const { apiKey, projectId, endpoint } = credentials(config);
  let response: Response;
  try {
    response = await fetch(apiUrl(endpoint, path), {
      method,
      headers: withoutUndefined({
        "Content-Type": "application/json",
        "x-api-key": apiKey,
        "x-project-id": projectId,
        Authorization: `Bearer ${apiKey}`,
        "Idempotency-Key": options.idempotencyKey,
      }) as Record<string, string>,
      body: options.body === undefined ? undefined : JSON.stringify(options.body),
    });
  } catch (error) {
    throw new ZrokyVerifiedActionError(
      `[ZROKY] verified action API unavailable: ${String((error as { message?: unknown })?.message ?? error)}`,
    );
  }
  if (!response.ok) {
    throw new ZrokyVerifiedActionError(`[ZROKY] verified action API failed with HTTP ${response.status}.`);
  }
  try {
    const payload = (await response.json()) as unknown;
    if (!isRecord(payload)) {
      throw new Error("non-object response");
    }
    return payload;
  } catch (error) {
    throw new ZrokyVerifiedActionError(
      `[ZROKY] verified action API returned invalid JSON: ${String((error as { message?: unknown })?.message ?? error)}`,
    );
  }
}

function requiresApproval(decision: JsonObject): boolean {
  return decision.requires_approval === true || decision.status === "approval_pending";
}

function blocked(decision: JsonObject): boolean {
  return decision.allowed === false && !requiresApproval(decision);
}

export async function verifiedAction(
  options: VerifiedActionOptions,
  config: ZrokyConfig = {},
): Promise<VerifiedActionDecision> {
  const resolved = resolveConfig(config);
  const executionRequest = validateExecutionRequest(options.executionRequest);
  const action = await requestJson(
    "POST",
    "/v1/action-intents",
    resolved,
    {
      idempotencyKey: options.idempotencyKey ?? defaultIdempotencyKey(),
      body: withoutUndefined({
        agent_id: options.agentId ?? resolved.agentId,
        contract_version: options.contractVersion,
        action_type: options.actionType,
        operation_kind: options.operationKind,
        environment: options.environment ?? "production",
        principal: options.principal ?? {},
        actor_chain: options.actorChain ?? [],
        purpose: options.purpose ?? {},
        resource: options.resource ?? {},
        parameters: options.parameters ?? {},
        execution_request: executionRequest,
        verification_profile: options.verificationProfile,
        deadline: deadlineValue(options.deadline),
        trace_context: options.traceContext,
      }),
    },
  );
  const actionId = text(action.action_id);
  if (!actionId) {
    throw new ZrokyVerifiedActionError("[ZROKY] verified action create response did not include action_id.", { action });
  }

  const decision = await requestJson("POST", `/v1/action-intents/${actionId}/decide`, resolved, { body: {} });
  if (requiresApproval(decision)) {
    if (options.raiseOnApproval === false) {
      return decision as VerifiedActionDecision;
    }
    throw new ZrokyVerifiedActionApprovalRequired(action, decision);
  }
  if (blocked(decision)) {
    throw new ZrokyVerifiedActionBlocked("[ZROKY] verified action was blocked by policy.", { action, decision });
  }
  return decision as VerifiedActionDecision;
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => {
    globalThis.setTimeout(resolve, ms);
  });
}

export async function awaitActionProof(
  actionId: string,
  options: AwaitActionProofOptions = {},
  config: ZrokyConfig = {},
): Promise<ActionProofResult> {
  const resolved = resolveConfig(config);
  const timeoutMs = Math.max(100, options.timeoutMs ?? 120_000);
  const pollIntervalMs = Math.max(50, options.pollIntervalMs ?? 2_000);
  const deadline = Date.now() + timeoutMs;
  let lastAction: JsonObject | undefined;

  while (Date.now() < deadline) {
    const action = await requestJson("GET", `/v1/action-intents/${actionId}`, resolved);
    lastAction = action;
    const proofStatus = String(action.proof_status ?? "");
    const receiptStatus = String(action.receipt_status ?? "");
    if (TERMINAL_PROOF_STATUSES.has(proofStatus) && TERMINAL_RECEIPT_STATUSES.has(receiptStatus)) {
      const receipt =
        receiptStatus === "generated"
          ? await requestJson("GET", `/v1/action-intents/${actionId}/receipt`, resolved)
          : null;
      return {
        actionId,
        action,
        receipt,
        proofStatus,
        receiptStatus,
        signatureValid: typeof receipt?.signature_valid === "boolean" ? receipt.signature_valid : null,
        evidenceId: typeof receipt?.receipt_id === "string" ? receipt.receipt_id : null,
      };
    }
    await delay(pollIntervalMs);
  }

  throw new ZrokyVerifiedActionError("[ZROKY] timed out waiting for verified action proof.", {
    action: lastAction,
  });
}
